DOC-557 | arangodump improved performance & resource usage limits (#295)
arangodb/docs-hugo: site/content/3.12/release-notes/version-3.12/whats-new-in-3-12.md

can open. Thus, these options should only be enabled on deployments with a
limited number of collections/shards/indexes.

## Client tools

### arangodump

#### Improved dump performance

> ArangoDB 3.12 includes extended parallelization capabilities to work not only

Reviewer: At the moment, it is tagged as an experimental feature, which I think we should mention here:
https://github.com/arangodb/arangodb/blob/4f154796969214eca5c25e143afb87fb9a16060d/client-tools/Dump/DumpFeature.cpp#L914

Reviewer 2: The "experimental" tagging will be removed by this upstream PR:
https://github.com/arangodb/arangodb/pull/20022

at the collection level, but also at the shard level. In combination with the
new optimized format, database dumps are now created and restored quickly and

> occupy minimal disk space. This major performance boost makes dumps five times

Reviewer: I think "more quickly" would be better, it wasn't horribly slow before.

What are the changes to the format that make the dumps smaller on disk?

Author: > What are the changes to the format that make the dumps smaller on disk?
@jsteemann maybe you can clarify this?

Reviewer 2: I also all for `more quickly` rather than `quickly`. The old dump variant wasn't too slow, as Simran already mentioned. Additionally, the new variant _can_ make dump & restore quicker than when using the old variant, it is not a guarantee that we are always 5 times faster than before. So this should be rephrased to "up to several times faster" instead of giving a precise number.

The potential speedup can be achieved by the following factors:
* new dump variant, enabled via `--use-parallel-dump true` (which now is also the default). The new variant uses prefetching and parallelization on the server, so that the server proactively keeps producing more results for arangodump to fetch. So that when an arangodump request comes in, the server can already respond with some ready-to-use result.
* optional: make arangodump write multiple output files per collection/shard. This can be enabled by setting the `--split-files` option to `true`. This is currently opt-in. The reason is that dumps that are created with this option enabled cannot be restored into previous versions of ArangoDB easily. The file splitting allows better parallelization when writing the results into the output file, which in case of non-split files must be serialized. The serialization can easily become a bottleneck especially when output files are gzip-compressed by arangodump.
* dumping the data into velocypack format instead of JSON. By setting the option `--dump-vpack`, the resulting dump data will be stored in velocypack format, not JSON. The velocypack format is normally more compact than JSON, so by using this option, the output file size can be reduced compared to JSON, even when compression is enabled. It can also lead to faster dumps, because less data needs to be shipped around and written. This is currently experimental and opt-in, for the reasons that only arangorestore from 3.12 or higher will be able to interpret and restore vpack dumps, and because there aren't many other tools than can read vpack data. So from the user side it may be unwanted to produce dumps in a different format that is not that much supported by other tools. But the option should be mentioned for users that want best dump performance and the smallest possible dumps.
* compressing the dump data on the server for transfer. By setting the option `--compress-transfer` to `true`, dump data can be compressed on the server for faster transfer. This is helpful especially if the network is slow or its capacity is maxed out. It won't make a difference otherwise.

Reviewer: @jsteemann I haven't looked at the dump output but if --dump-vpack is used, can the arangovpack tool be used to convert (parts of) the dump from VPack to JSON, and does this work with reasonable speed and memory usage? If not, then we might want to create a ticket for that.

Reviewer 2: Unfortunately the arangovpack tool **cannot** be used for that.
The reason is that arangodump with `--dump-vpack` produces individual batches of data, which are all valid velocypack, but these batches are all written to the same file, one after the other.
Currently arangovpack will either only handle the velocypack from the first batch of data, or even fail. I am not sure because I didn't test that. 
But I am sure we would need to augment arangovpack first. It is probably a good idea to do it anyway, but it hasn't been done yet.

Reviewer: https://arangodb.atlassian.net/browse/FRB-340

faster and restores three times faster, which is extremely useful when dealing

> with large shards.

Reviewer: Is the `--parallel-dump` option that controls whether the new dumping implementation is used not mentioned on purpose? The default appears to be false (option disabled):
https://github.com/arangodb/arangodb/blob/4f154796969214eca5c25e143afb87fb9a16060d/client-tools/Dump/DumpFeature.h#L86

Reviewer 2: Correct, it was indeed still `false` by default.
I haven't had time to make it the default. But now there is upstream PR https://github.com/arangodb/arangodb/pull/20022, which should make it the default.

can open. Thus, these options should only be enabled on deployments with a
limited number of collections/shards/indexes.

## Client tools

### arangodump

#### Improved dump performance

ArangoDB 3.12 includes extended parallelization capabilities to work not only
at the collection level, but also at the shard level. In combination with the
new optimized format, database dumps are now created and restored quickly and
occupy minimal disk space. This major performance boost makes dumps five times
faster and restores three times faster, which is extremely useful when dealing
with large shards.

#### Resource usage limits

The following startup options that can be used to limit

> the resource usage of parallel _arangodump_ invocations have been added:

Reviewer 2: It should be mentioned here explicitly that these are startup options for arangod, not for arangodump.

can open. Thus, these options should only be enabled on deployments with a
limited number of collections/shards/indexes.

## Client tools

### arangodump

> #### Improved dump performance

Reviewer: ```suggestion
#### Improved dump performance and size
```

ArangoDB 3.12 includes extended parallelization capabilities to work not only
at the collection level, but also at the shard level. In combination with the

> new optimized format, database dumps are now created and restored more quickly

Reviewer: ```suggestion
new VelocyPack format, database dumps are now created and restored more quickly
```

> and occupy minimal disk space. This major performance boost makes dumps and

Reviewer: ```suggestion
and occupy less disk space. This major performance boost makes dumps and
```

restores up to several times faster, which is extremely useful when dealing
with large shards.

> The new dump variant can be enabled via `--use-parallel-dump`. The default
> value is `true`.

Reviewer: ```suggestion
Whether the new parallel dump variant is used is controlled by the newly added
`--use-parallel-dump` startup option. The default value is `true`.
```

To achieve the best dump performance and the smallest data dumps in terms of

> size, you can use the `--dump-vpack` option. The resulting dump data is stored

Reviewer: ```suggestion
size, you can additionally use the `--dump-vpack` option. The resulting dump data is stored
```

> in velocypack format instead of JSON. The velocypack format is more compact than

Reviewer: ```suggestion
in VelocyPack format instead of JSON. The VelocyPack format is more compact than
```

JSON, therefore the output file size can be reduced compared to JSON, even

> when compression is enabled, and can also lead to faster dumps. Note, however,
> that this option is experimental and disabled by default.

Reviewer: ```suggestion
when compression is enabled. It can also lead to faster dumps because there is less data to
transfer and no conversion from the server-internal format (VelocyPack) to JSON is needed.
Note, however, that this option is experimental and disabled by default.
```

Optionally, you can make _arangodump_ write multiple output files per
collection/shard. The file splitting allows better parallelization when
writing the results into the output file, which in case of non-split files
must be serialized.
You can enable it by setting the `--split-files` option to `true`. This option
is disabled by default considering that dumps created with this option enabled

> cannot be restored into previous versions of ArangoDB easily.

Reviewer: ```suggestion
cannot be restored into previous versions of ArangoDB.
```
